File: VRSampleApp/VRSampleApp/Classes/PresentationLayer/Modules/ListOfVideosModule/ViewModel/ListOfVideosFactory.m
//
//  ListOfVideosFactory.m
//  VRSampleApp
//
//  Copyright © 2017 Ooyala Inc. All rights reserved.
//

#import "ListOfVideosFactory.h"
#import "DefaultVideoPlayerViewController.h"
#import "IMAVideoPlayerViewController.h"
#import "VideoPlayerViewModel.h"
#import "CustomVideoViewController.h"
#import "CustomVideoViewModel.h"
#import "VideoItem.h"

@implementation ListOfVideosFactory

#pragma mark - Public functions

- (UIViewController *)viewControllerWithVideoItem:(VideoItem *)videoItem
                                            pcode:(NSString *)pcode
                                           domain:(NSString *)domain
                                    QAModeEnabled:(BOOL)QAModeEnabled {
  UIViewController *configuredViewController;

  switch (videoItem.videoAdType) {
    case IMA:
      configuredViewController = [self createIMAViewPlayerModule:videoItem
                                                           pcode:pcode
                                                          domain:domain
                                                andQAModeEnabled:QAModeEnabled];
      break;
      
    default:
      configuredViewController = [self createDefaultVideoPlayerModuleWithVideoItem:videoItem
                                                                             pcode:pcode
                                                                            domain:domain
                                                                  andQAModeEnabled:QAModeEnabled];
      break;
  }
  
  return configuredViewController;
}

- (UIViewController *)configuredCustomVideoViewControllerWithCompletion:(void (^)(NSString *pCode, NSString *embedCode))completion {
  CustomVideoViewModel *videoPlayerViewModel = [CustomVideoViewModel new];
  CustomVideoViewController *customVideoViewController = [[UIStoryboard storyboardWithName:@"Main" bundle:nil] instantiateViewControllerWithIdentifier:@"CustomVideoViewController"];
  
  customVideoViewController.viewModel = videoPlayerViewModel;
  videoPlayerViewModel.delegate = customVideoViewController;
  videoPlayerViewModel.completionBlock = completion;
  
  return customVideoViewController;
}

#pragma mark - Private functions

- (UIViewController *)createDefaultVideoPlayerModuleWithVideoItem:(VideoItem *)videoItem
                                                            pcode:(NSString *)pcode
                                                           domain:(NSString *)domain
                                                 andQAModeEnabled:(BOOL)QAModeEnabled {
  
  VideoPlayerViewModel *videoPlayerViewModel = [[VideoPlayerViewModel alloc] initWithVideoItem:videoItem
                                                                                         pcode:pcode
                                                                                        domain:domain
                                                                              andQAModeEnabled:QAModeEnabled];

  DefaultVideoPlayerViewController *viewController = [[UIStoryboard storyboardWithName:@"Main" bundle:nil] instantiateViewControllerWithIdentifier:@"DefaultVideoPlayerViewController"];
  
  viewController.viewModel = videoPlayerViewModel;

  return viewController;
}

- (UIViewController *)createIMAViewPlayerModule:(VideoItem *)videoItem
                                                            pcode:(NSString *)pcode
                                                           domain:(NSString *)domain
                                                 andQAModeEnabled:(BOOL)QAModeEnabled {
  
  VideoPlayerViewModel *videoPlayerViewModel = [[VideoPlayerViewModel alloc] initWithVideoItem:videoItem
                                                                                         pcode:pcode
                                                                                        domain:domain
                                                                              andQAModeEnabled:QAModeEnabled];

  IMAVideoPlayerViewController *viewController = [[UIStoryboard storyboardWithName:@"Main" bundle:nil] instantiateViewControllerWithIdentifier:@"IMAVideoPlayerViewController"];
  
  viewController.viewModel = videoPlayerViewModel;
  
  return viewController;
}

@end

File: SwiftVRSkinSampleApp/SwiftVRSkinSampleApp/Classes/PresentationLayer/Modules/ListOfVideosModule/ViewModel/ListOfVideosFactory.swift
//
//  ListOfVideosFactory.swift
//  VRSkinSampleApp
//
//  Copyright © 2017 Ooyala, Inc. All rights reserved.
//

import UIKit

class ListOfVideosFactory {
  // MARK: - Public functions
  func viewController(withVideoItem: VideoItem,
                      pcode: String,
                      domain: String,
                      QAModeEnabled: Bool) -> UIViewController {
    var configuredViewController: UIViewController

    switch withVideoItem.videoAdType {
    case .ima:
      configuredViewController = createIMAViewPlayerModule(withVideoItem: withVideoItem,
                                                           pcode: pcode,
                                                           domain: domain,
                                                           QAModeEnabled: QAModeEnabled)
    default:
      configuredViewController = createDefaultVideoPlayerModule(withVideoItem: withVideoItem,
                                                                pcode: pcode,
                                                                domain: domain,
                                                                QAModeEnabled: QAModeEnabled)
    }

    return configuredViewController
  }

  func configuredCustomVideoViewController(with
    completion: ((_ pCode: String, _ embedCode: String) -> Void)?) -> UIViewController {

    let mainStoryboard = UIStoryboard(name: "Main", bundle: nil)
    guard let customVideoViewController =
      mainStoryboard.instantiateViewController(withIdentifier: "CustomVideoViewController")
        as? CustomVideoViewController else {
        return UIViewController()
    }

    let customVideoViewModel = CustomVideoViewModel()
    customVideoViewController.viewModel = customVideoViewModel
    customVideoViewModel.delegate = customVideoViewController
    customVideoViewModel.completionBlock = completion

    return customVideoViewController
  }

  // MARK: - Private functions
  private func createDefaultVideoPlayerModule(withVideoItem: VideoItem,
                                              pcode: String,
                                              domain: String,
                                              QAModeEnabled: Bool) -> UIViewController {
    let videoPlayerViewModel = VideoPlayerViewModel(videoItem: withVideoItem,
                                                    pcode: pcode,
                                                    domain: domain,
                                                    QAModeEnabled: QAModeEnabled)
    let mainStoryboard = UIStoryboard(name: "Main", bundle: nil)
    guard let videoPlayerViewController =
      mainStoryboard.instantiateViewController(withIdentifier: "DefaultVideoPlayerViewController")
        as? DefaultVideoPlayerViewController else {
      return UIViewController()
    }

    videoPlayerViewController.viewModel = videoPlayerViewModel

    return videoPlayerViewController
  }

  private func createIMAViewPlayerModule(withVideoItem: VideoItem,
                                         pcode: String,
                                         domain: String,
                                         QAModeEnabled: Bool) -> UIViewController {
    let videoPlayerViewModel = VideoPlayerViewModel(videoItem: withVideoItem,
                                                    pcode: pcode,
                                                    domain: domain,
                                                    QAModeEnabled: QAModeEnabled)
    let mainStoryboard = UIStoryboard(name: "Main", bundle: nil)
    guard let videoPlayerViewController =
      mainStoryboard.instantiateViewController(withIdentifier: "IMAVideoPlayerViewController")
        as? IMAVideoPlayerViewController else {
      return UIViewController()
    }

    videoPlayerViewController.viewModel = videoPlayerViewModel

    return videoPlayerViewController
  }
}
